consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+                    None => self
+                        .db_tx
+                        .get_mainchain_tx_index(&tx.get_id())?
+                        .ok_or(BlockError::Unknown)?,
+                };
+
+                if input_index >= tx_index.get_output_count() {
+                    return Err(BlockError::Unknown);
+                }
+                tx_index.unspend(input_index).map_err(BlockError::from)?;
+                cached_inputs.insert(tx.get_id(), tx_index.clone());
+            }
+            // TODO: Seems, would be better to remove TxMainChainIndex for disconnected block.
+            // self.db_tx.del_mainchain_tx_index(&tx.get_id())?;
+        }
+        self.store_cached_inputs(&cached_inputs)?;
+        Ok(())
+    }
+
+    fn get_mainchain_index_by_outpoint(
+        tx_db: &<blockchain_storage::Store as Transactional<'a>>::TransactionRw,
+        outpoint: &OutPoint,
+    ) -> Result<TxMainChainIndex, BlockError> {
+        let tx_id = match outpoint.get_tx_id() {
+            OutPointSourceId::Transaction(tx_id) => tx_id,
+            OutPointSourceId::BlockReward(_) => {
+                unimplemented!()
+            }
+        };
+        tx_db.get_mainchain_tx_index(&tx_id)?.ok_or(BlockError::Unknown)
+    }
+
+    fn get_tx_by_outpoint(
+        tx_db: &<blockchain_storage::Store as Transactional<'a>>::TransactionRw,
+        outpoint: &OutPoint,
+    ) -> Result<Transaction, BlockError> {
+        let tx_id = match outpoint.get_tx_id() {
+            OutPointSourceId::Transaction(tx_id) => tx_id,
+            OutPointSourceId::BlockReward(_) => {
+                unimplemented!()
+            }
+        };
+        let tx_index = tx_db.get_mainchain_tx_index(&tx_id)?.ok_or(BlockError::Unknown)?;
+        match tx_index.get_tx_position() {
+            SpendablePosition::Transaction(position) => {
+                tx_db.get_mainchain_tx_by_position(position)?.ok_or(BlockError::Unknown)
+            }
+            SpendablePosition::BlockReward(_) => unimplemented!(),
+        }
+    }
+
+    fn get_output_by_outpoint(
+        tx_db: &<blockchain_storage::Store as Transactional<'a>>::TransactionRw,
+        outpoint: &OutPoint,
+    ) -> Result<common::chain::TxOutput, BlockError> {
+        let tx = Self::get_tx_by_outpoint(tx_db, outpoint)?;
+        let output_index = outpoint.get_output_index() as usize;
+        assert!(output_index <= tx.get_outputs().len());
+        Ok(tx.get_outputs()[output_index].clone())
+    }
+
+    fn get_input_value(
+        tx_db: &<blockchain_storage::Store as Transactional<'a>>::TransactionRw,
+        input: &common::chain::TxInput,
+    ) -> Result<Amount, BlockError> {
+        let tx = Self::get_tx_by_outpoint(tx_db, input.get_outpoint())?;
+        let output_index = input.get_outpoint().get_output_index() as usize;
+        assert!(output_index <= tx.get_outputs().len());
+        Ok(tx.get_outputs()[output_index].get_value())
```

Comment:
Same here, is the check correct?

Reply:
Also, Anton, please, NEVER use the [] operator... never, and I'm not sure whether there are exceptions. This isn't C++. There are many reasons for this. First, the [] operator is checked. So you're double-checking the range. Second, if it's wrong, it'l panic. So, you should do it the right way where you get() and match for that.

In C++ it's the opposite. The operator[] is unchecked and at() is checked with an exception if out of range.

Whenever you use [], think whether you're doing the right thing.

Reply:
Ohh, thanks! So, fixed places where [] operator was being used. @altonen related to your previous message, I fixed a couple of if-conditions where might taking item out of array bounds.